### ISSUE TITLE
Performance regression in text shaper

### ISSUE THREAD

**Issue by Jacalz** [fyne-io/fyne#4008]
### Checklist

- [X] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same problem, before opening a new one.
- [X] This issue only relates to a single bug. I will open new issues for any other problems.

### Describe the bug

During review of https://github.com/fyne-io/fyne/pull/4007, it was noticed that the handling of carriage return characters on `develop` has regressed since v2.3.x. The change is that we no longer ignore carriage returns while iterating over the characters but rather remove all carriage return characters in whole input string before starting shaping.

The impact of this change is worse performance for all paths and then potentially large heap allocations in the cases where the text contains any `\r` characters. For text that doesn't, the slowdown is simply an O(n) iteration over the string to see if it contains any of those characters. No allocations, but allocating over the string twice is unnecessary work done, especially for long strings if it happens often. If the text does contain those characters, the whole string is re-allocated with the characters removed. That will be very slow and produce memory usage increases due to the large heap allocations.

Previous code: https://github.com/fyne-io/fyne/blob/1834d704a305f4729cce65213f5b5c3098e1d993/internal/painter/font.go#L260
New code: https://github.com/fyne-io/fyne/blob/03384a7a3aa88f328724e399301ad38b321dcd84/internal/painter/font.go#L159

### How to reproduce

Use long text containing carriage return characters. This issue is more from a code review viewpoint. 

### Example code

This is about the change in code. See the comparison below:
- Previous code: https://github.com/fyne-io/fyne/blob/1834d704a305f4729cce65213f5b5c3098e1d993/internal/painter/font.go#L260
- New code: https://github.com/fyne-io/fyne/blob/03384a7a3aa88f328724e399301ad38b321dcd84/internal/painter/font.go#L159

### Fyne version

03384a7a3aa88f328724e399301ad38b321dcd84

### Go compiler version

1.20.5

### Operating system and version

Fedora Linux 38

**Comment by andydotxyz** [comment]
The question is what is correct here. The slower code gets the kerning right (it shapes the characters before and after the ignored item) whereas before we were not shaping at all so just pushed the next character into the line.

I think the visuals we have now are better and changes may have to be made up in the shaper module.

I'm not sure I see this as a release blocker - is there a concept of just how much or a degradation in performance this is?

**Comment by Jacalz** [comment]
Yeah, blocker was probably a bit of an exaggeration. I'm not saying that the previous output was better just that we ought to fix this performance penalty.

I had a second look at the code and now see why it will require changes in the shaper (I missed that earlier, thought that the string only was used inside the loop).

Will have a look at the performance one of the coming days but I think this is something we will want to resolve sooner or later either way.

**Comment by andydotxyz** [comment]
That sounds like a plan :)

**Comment by andydotxyz** [comment]
Big improvements underway and already landed too, I think this is resolved
